tree/tree.go

```diff
@@ -0,0 +1,160 @@
+package tree
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/vocdoni/arbo"
+	"go.vocdoni.io/dvote/db"
+	"go.vocdoni.io/dvote/db/badgerdb"
+)
+
+// Tree defines the struct that implements the MerkleTree functionalities
+type Tree struct {
+	t  *arbo.Tree
+	db db.Database
+}
+
+// Options is used to pass the parameters to load a new Tree
+type Options struct {
+	// Path of the db.Database to be used
+	Path string
+	// MaxLevels that will have the Tree
+	MaxLevels int
+	// HashFunc defines the hash function that the tree will use
+	HashFunc arbo.HashFunction
+}
+
+// New returns a new Tree, if there already is a Tree in the database, it will
+// load it
+func New(opts Options) (*Tree, error) {
+	database, err := badgerdb.New(badgerdb.Options{Path: opts.Path})
+	if err != nil {
+		return nil, err
+	}
+
+	tree, err := arbo.NewTree(database, opts.MaxLevels, opts.HashFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Tree{
+		t:  tree,
+		db: database,
+	}, nil
+}
+
+func (t *Tree) DB() db.Database {
+	return t.db
+}
+
+// Get returns the value for a given key.
+func (t *Tree) Get(rTx db.ReadTx, key []byte) ([]byte, error) {
+	_, v, err := t.t.GetWithTx(rTx, key)
+	if err != nil {
+		return nil, err
+	}
+
+	return v, nil
+}
+
+// Set adds or updates a key. First performs a Get, and if the key does not
+// exist yet it will call Add, while if the key already exists, it will perform
+// an Update. If the non-existence or existence of the key is already known, is
+// more efficient to directly call Add or Update.
+func (t *Tree) Set(wTx db.WriteTx, key, value []byte) error {
+	_, err := t.Get(wTx, key)
+	if err == arbo.ErrKeyNotFound {
+		// key does not exist, use Add
+		return t.Add(wTx, key, value)
+	} else if err != nil {
+		return err
+	}
+	// key already exists, use Update
+	return t.t.UpdateWithTx(wTx, key, value)
+}
+
+// Add adds a new leaf, if the key already exists will return error
+func (t *Tree) Add(wTx db.WriteTx, key, value []byte) error {
+	return t.t.AddWithTx(wTx, key, value)
+}
+
+// AddBatch adds a batch of key-values to the Tree. Returns an array containing
+// the indexes of the keys failed to add. Supports empty values as input
+// parameters, which is equivalent to 0 valued byte array.
+func (t *Tree) AddBatch(wTx db.WriteTx, keys, values [][]byte) ([]int, error) {
+	return t.t.AddBatchWithTx(wTx, keys, values)
+}
+
+func (t *Tree) Iterate(callback func(key, value []byte) bool) error {
+	callbackWrapper := func(currLvl int, k, v []byte) bool {
+		callback(k, v)
+		return true
```

Comment:
```suggestion
		return callback(k, v)
```
Otherwise you're ignoring the callback return value, which means the iteration can't be stopped in the middle!

Reply:
Right! Updated